src/client/opamConfigCommand.ml

```diff
@@ -518,26 +521,38 @@ let switch_allowed_fields, switch_allowed_sections =
   (fun () -> Lazy.force allowed_fields),
   fun () -> Lazy.force allowed_sections
 
-let confset_switch st =
-  let root = st.switch_global.root in
-  let config_f = OpamPath.Switch.switch_config root st.switch in
+let confset_switch gt switch switch_config =
+  let config_f = OpamPath.Switch.switch_config gt.root switch in
   let write new_config = OpamFile.Switch_config.write config_f new_config in
   { stg_fields = OpamFile.Switch_config.fields;
     stg_allwd_fields = switch_allowed_fields ();
     stg_sections = OpamFile.Switch_config.sections;
     stg_allwd_sections = switch_allowed_sections ();
-    stg_config = st.switch_config;
+    stg_config = switch_config;
     stg_write_config = write;
-    stg_doc = switch_doc st
+    stg_doc = switch_doc switch
   }
 
-let set_opt_switch_t ?inner st field value =
-  let switch_config =
-    set_opt ?inner field value (confset_switch st)
-  in
-  { st with switch_config }
-
-let set_opt_switch = set_opt_switch_t ~inner:false
+let with_switch lock gt st k =
```

Comment:
I would add a type annotation here to clarify the lock-kind of the switch state (might be necessary since we use phantom types, e.g. to ensure the consistency of the required lock and provided st). It would also clarify that the `st` argument is actually an `st_opt`.
```suggestion
let with_switch:
  'a lock -> 'b global_state -> 'a switch_state option
  -> (switch -> OpamFile.Switch_config.t -> 'c) -> 'c =
  fun lock gt st k ->
```